harvester_e2e_tests/integration/test_vm_functions.py

```diff
@@ -466,3 +507,642 @@ def test_delete(self, api_client, unique_vm_name, wait_timeout):
+            )
+
+        # Remove cloned VM and volumes
+        code, data = api_client.vms.get(cloned_name)
+        cloned_spec = api_client.vms.Spec.from_dict(data)
+        api_client.vms.delete(cloned_name)
+        endtime = datetime.now() + timedelta(seconds=wait_timeout)
+        while endtime > datetime.now():
+            code, data = api_client.vms.get(cloned_name)
+            if 404 == code:
+                break
+            sleep(3)
+        else:
+            raise AssertionError(
+                f"Failed to Delete VM({cloned_name}) with errors:\n"
+                f"Status({code}): {data}"
+            )
+        for vol in cloned_spec.volumes:
+            vol_name = vol['volume']['persistentVolumeClaim']['claimName']
+            api_client.volumes.delete(vol_name)
+
+    def test_clone_stopped_vm(self, api_client, ssh_keypair, wait_timeout,
+                              host_shell, vm_shell, stopped_vm):
+        """
+        To cover test:
+        - (legacy) https://harvester.github.io/tests/manual/virtual-machines/clone-vm-that-is-turned-off/ # noqa
+        - (new) https://github.com/harvester/tests/issues/361
+
+        Steps:
+            1. Create a VM with 1 CPU 2 Memory
+            2. Start the VM and write some data
+            3. Stop the VM
+            4. Clone the VM into VM-cloned
+            5. Verify VM-Cloned
+
+        Exepected Result:
+            - Cloned-VM should be available and stopped
+            - Cloned-VM should able to start and becomes `Running`
+            - Written data should available in Cloned-VM
+        """
+        unique_vm_name, ssh_user = stopped_vm
+        pub_key, pri_key = ssh_keypair
+        code, data = api_client.vms.start(unique_vm_name)
+
+        endtime = datetime.now() + timedelta(seconds=wait_timeout)
+        while endtime > datetime.now():
+            code, data = api_client.vms.get_status(unique_vm_name)
+            if 200 == code:
+                phase = data.get('status', {}).get('phase')
+                conds = data.get('status', {}).get('conditions', [{}])
+                if ("Running" == phase
+                   and "AgentConnected" == conds[-1].get('type')
+                   and data['status'].get('interfaces')):
+                    break
+            sleep(3)
+        else:
+            raise AssertionError(
+                f"Failed to Start VM({unique_vm_name}) with errors:\n"
+                f"Status: {data.get('status')}\n"
+                f"API Status({code}): {data}"
+            )
+        vm_ip = next(iface['ipAddress'] for iface in data['status']['interfaces']
+                     if iface['name'] == 'default')
+        code, data = api_client.hosts.get(data['status']['nodeName'])
+        host_ip = next(addr['address'] for addr in data['status']['addresses']
+                       if addr['type'] == 'InternalIP')
+
+        # Log into VM to make some data
+        with host_shell.login(host_ip, jumphost=True) as h:
```

Comment:
when we open the context manager maybe we could name it something other than `h`?  I wonder if tox pep8 would argue about the line length if we named it something like `hstshell`?

Reply:
I think short variable name is one of conventions in Python, like the example[^1] of `open`:
```python
import os
dir_fd = os.open('somedir', os.O_RDONLY)
def opener(path, flags):
    return os.open(path, flags, dir_fd=dir_fd)

with open('spamspam.txt', 'w', opener=opener) as f:
    print('This will be written to somedir/spamspam.txt', file=f)

os.close(dir_fd)  # don't leak a file descriptor
```
Usually we will let variable name short as its living time (not scope), so sometimes you would able find some code snippets like:
```python
# https://github.com/aio-libs/aiohttp/blob/3ff5fde3051f552a0aa2ac3678ab7f5fbdc007e1/aiohttp/web.py#L328
for h in host:
    sites.append(TCPSite(...))
```
and emphasize the variable by renaming it:
https://github.com/harvester/tests/blob/4baf4bc1531c49ff561bdedf63f1b7785f8c61dc/harvester_e2e_tests/integration/test_vm_functions.py#L601-L607

It would be a bit conflict with the suggestion: meaningful variable name (one of the topic in _clean code_, targeting for Java), but yeah, it should be different, because Java don't have any JEP like PEP8.

[^1]: https://docs.python.org/3/library/functions.html#open

Reply:
That's a great mention :smile:  - it sounds good to leave it short.
I've seen some context managers as:
```python
with open('spamspam.txt', 'w', opener=opener) as file:
    print('context manager open... with file....')
```
Where `file` was used instead of `f` but they both mean the same thing and the shorter works great :smile: .